pkg/ccl/streamingccl/streamproducer/event_stream.go

```diff
@@ -255,97 +199,85 @@ func (s *eventStream) Close(ctx context.Context) {
-	case <-ctx.Done():
-	case s.eventsCh <- kvcoord.RangeFeedMessage{
-		RangeFeedEvent: &kvpb.RangeFeedEvent{SST: sst},
-		RegisteredSpan: registeredSpan,
-	}:
-		log.VInfof(ctx, 1, "onSSTable: %s@%s with registered span %s",
-			sst.Span, sst.WriteTS, registeredSpan)
+	if err := s.addSST(sst, registeredSpan); err != nil {
+		s.errCh <- err
+		return
+	}
+	if err := s.maybeFlushBatch(ctx); err != nil {
+		s.errCh <- err
 	}
 }
 
 func (s *eventStream) onDeleteRange(ctx context.Context, delRange *kvpb.RangeFeedDeleteRange) {
-	select {
-	case <-ctx.Done():
-	case s.eventsCh <- kvcoord.RangeFeedMessage{RangeFeedEvent: &kvpb.RangeFeedEvent{DeleteRange: delRange}}:
-		log.VInfof(ctx, 1, "onDeleteRange: %s@%s", delRange.Span, delRange.Timestamp)
+	s.seb.addDelRange(*delRange)
+	if err := s.maybeFlushBatch(ctx); err != nil {
+		s.errCh <- err
 	}
 }
 
-// makeCheckpoint generates checkpoint based on the frontier.
-func makeCheckpoint(f span.Frontier) (checkpoint streampb.StreamEvent_StreamCheckpoint) {
-	f.Entries(func(sp roachpb.Span, ts hlc.Timestamp) (done span.OpResult) {
-		checkpoint.ResolvedSpans = append(checkpoint.ResolvedSpans, jobspb.ResolvedSpan{
-			Span:      sp,
-			Timestamp: ts,
-		})
+func (s *eventStream) shouldCheckpoint(ctx context.Context, advanced bool) bool {
+	age := timeutil.Since(s.lastCheckpointTime)
+	return (advanced && age > s.spec.Config.MinCheckpointFrequency) || (age > 2*s.spec.Config.MinCheckpointFrequency)
+}
+
+func (s *eventStream) sendCheckpoint(ctx context.Context, frontier rangefeed.VisitableFrontier) {
+	if err := s.flushBatch(ctx); err != nil {
+		s.errCh <- err
+		return
+	}
+
+	spans := make([]jobspb.ResolvedSpan, 0, s.lastCheckpointLen)
+	frontier.Entries(func(sp roachpb.Span, ts hlc.Timestamp) (done span.OpResult) {
+		spans = append(spans, jobspb.ResolvedSpan{Span: sp, Timestamp: ts})
 		return span.ContinueMatch
 	})
-	return
+	s.lastCheckpointLen = len(spans)
+
+	if err := s.sendFlush(ctx, &streampb.StreamEvent{Checkpoint: &streampb.StreamEvent_StreamCheckpoint{ResolvedSpans: spans}}); err != nil {
+		s.errCh <- err
+		return
+	}
+	// set the local time for pacing.
+	s.lastCheckpointTime = timeutil.Now()
+}
+
+func (s *eventStream) maybeFlushBatch(ctx context.Context) error {
+	if s.seb.size > int(s.spec.Config.BatchByteSize) {
+		return s.flushBatch(ctx)
+	}
+	return nil
 }
 
-func (s *eventStream) flushEvent(ctx context.Context, event *streampb.StreamEvent) error {
+func (s *eventStream) flushBatch(ctx context.Context) error {
+	if s.seb.size == 0 {
+		return nil
+	}
+
+	defer s.seb.reset()
+	return s.sendFlush(ctx, &streampb.StreamEvent{Batch: &s.seb.batch})
+}
+func (s *eventStream) sendFlush(ctx context.Context, event *streampb.StreamEvent) error {
 	data, err := protoutil.Marshal(event)
 	if err != nil {
 		return err
 	}
-
 	select {
 	case <-ctx.Done():
 		return ctx.Err()
 	case s.streamCh <- tree.Datums{tree.NewDBytes(tree.DBytes(data))}:
 		return nil
-	case <-s.doneChan:
-		return nil
```

Comment:
I'm assuming we can get rid of this because the rangefeed library will cancel the context when we close it?

Reply:
indeed